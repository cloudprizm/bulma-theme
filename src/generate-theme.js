const fs = require('fs')
const path = require('path')
const findNodeModules = require('find-node-modules')

const { enableRuntimeTranspilation } = require('@hungry/babel-preset-cli')
enableRuntimeTranspilation()

const { parseFile } = require('@hungry/sass-ts-theme-creator')

const withinSRC = (file) => path.resolve(__dirname, file)

const typescriptThemeFunction = parseFile({
  file: withinSRC('./variables.sass'),
  includePaths: findNodeModules({ cwd: __dirname }).map(withinSRC)
})

fs.writeFileSync(withinSRC('__generated__/theme.ts'), typescriptThemeFunction)