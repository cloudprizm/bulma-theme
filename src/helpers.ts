import makeColor from 'color'

const toHex = (colorString: string) =>
  color(colorString).hex()

const isNestedColor = (val: string) =>
  val.indexOf('rgba') > -1 && val.indexOf('#') > -1

const resolveNestedHEXColor = (val: string) =>
  val.replace(/\#[0-9a-fA-F]+/,
    innerValue => color(innerValue).array().join(',')
  )

export const color = (val: string | string[]) =>
  makeColor(val)

export const toColorString = (val: string) =>
  isNestedColor(val)
    ? resolveNestedHEXColor(val)
    : toHex(val)

export const invertColor = (colorString: string) =>
  color(colorString).luminosity() > 0.55
    ? toColorString('rgba(0, 0, 0, 0.7)')
    : toColorString('#fff')

const isList = (d: string[]) =>
  d.filter(c => c === ',').length === (d.filter(c => c !== ',').length - 1)

const dimensionExtension = [/px/, /em/]
const hasDimension = (d: string[]) =>
  d.filter(c => dimensionExtension.map(p => p.test(c)).length > 0).length > 0

const hasOperator = (d: string[]) =>
  d.filter(c => ['+', '-', '*', '/'].indexOf(c) > -1).length > 0

const hasColorToResolve = (d: string[]) =>
  d.filter(isNestedColor).length > 0

const resolveValuesWithColor = (d: string[]) =>
  d.map(i => isNestedColor(i) ? toColorString(i) : i).join(' ')

const listJoin = (d: string[]) => d.join('')
const resolveDimensionValue = (d: string[]) => {
  const dimensions = dimensionExtension
    .map(e => e.test(d.join(' ')) ? e : null)
    .filter(Boolean)

  if (dimensions.length > 1)
    throw new Error(`Not able to resolve mixed dimensions, for ${d.join(' ')}`)

  const dimension = dimensions[0] as RegExp
  const dimensionReplacer = new RegExp(dimension as RegExp, 'g')
  const dimensionString = dimension.toString().replace(/\//g, '')
  const fn = d.join(' ').replace(dimensionReplacer, '')
  const safeEval = new Function(`return ${fn}`)
  return `${safeEval()}${dimensionString}`
}

// drop this ugly IF below
// const arrayResolvers = [
//   [isList, listJoin],
//   [and(hasOperator, hasDimension), resolveDimensionValue]
//   [hasColorToResolve, resolveValuesWithColor]
// ]

export const evaluate = (val: string | string[], variableName?: string) => {
  if (!Array.isArray(val)) return val

  if (isList(val)) return listJoin(val)
  if (hasOperator(val) && hasDimension(val)) return resolveDimensionValue(val)
  if (hasColorToResolve(val)) return resolveValuesWithColor(val)

  return val.join(' ')
}
