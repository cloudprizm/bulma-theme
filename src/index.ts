
export * from './runnable-theme'
export * from './helpers'
export * from './__generated__/theme'
