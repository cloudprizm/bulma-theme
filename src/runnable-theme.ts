import { toColorString, evaluate, invertColor } from './helpers'
import { makeBasicTheme, InjectedMethods, BulmaTheme } from './default.theme'
import { reader, asks, Reader } from 'fp-ts/lib/Reader'

export const resolvers: InjectedMethods = {
  color: toColorString,
  findColorInvert: invertColor,
  evaluate,
}
export type RunnableTheme = Reader<InjectedMethods, BulmaTheme>

export const theme: Reader<InjectedMethods, BulmaTheme> =
  reader
    .of<InjectedMethods, Partial<BulmaTheme>>({})
    .chain(input => asks(methods => makeBasicTheme(methods)(input)))

export const defaultTheme = theme.run(resolvers)

export const zLayers = {
  dropdown: defaultTheme.dropdownContentZ,
  modal: defaultTheme.modalZ,
  navbarDropdown: defaultTheme.navbarDropdownZ,
  navbarFixed: defaultTheme.navbarFixedZ,
  navbar: defaultTheme.navbarZ,
}
export type ThemeOverriding = (theme: BulmaTheme) => BulmaTheme