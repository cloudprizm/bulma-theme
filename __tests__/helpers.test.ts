import { evaluate } from '../src/helpers'

test('eval box shadow with colors', () => {
  const black = '#000'
  const input = [
    `0`,
    `8px`,
    `8px`,
    `rgba(${black}, 0.1)`,
    `,`,
    `0`,
    `0`,
    `0`,
    `1px`,
    `rgba(${black}, 0.1)`
  ]

  const boxShadowValue = evaluate(input, 'boxShadow')
  expect(boxShadowValue).toMatchSnapshot()
})

test('single negative value', () => {
  const input = [`-4px`]
  const result = evaluate(input)
  expect(result).toBe('-4px')
})

test('resolve px values', () => {
  const gap = '10px'
  const input = [`960px`, `+`, `(2 * ${gap})`]
  const desktopWidth = evaluate(input, 'desktop')
  expect(desktopWidth).toBe('980px')
})

test('resolve font family', () => {
  const input = [
    `BlinkMacSystemFont`, `,`,
    `-apple-system`, `,`,
    `"Segoe UI"`, `,`,
    `"Roboto"`, `,`,
    `"Oxygen"`, `,`,
    `"Ubuntu"`, `,`,
    `"Cantarell"`, `,`,
    `"Fira Sans"`, `,`,
    `"Droid Sans"`, `,`,
    `"Helvetica Neue"`, `,`,
    `"Helvetica"`, `,`,
    `"Arial"`, `,`,
    `sans-serif`
  ]

  const fontValue = evaluate(input, 'familySansSerif')
  expect(fontValue).toMatchSnapshot()
})