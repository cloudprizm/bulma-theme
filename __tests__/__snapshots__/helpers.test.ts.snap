// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eval box shadow with colors 1`] = `"0 8px 8px rgba(0,0,0, 0.1) , 0 0 0 1px rgba(0,0,0, 0.1)"`;

exports[`resolve font family 1`] = `"BlinkMacSystemFont,-apple-system,\\"Segoe UI\\",\\"Roboto\\",\\"Oxygen\\",\\"Ubuntu\\",\\"Cantarell\\",\\"Fira Sans\\",\\"Droid Sans\\",\\"Helvetica Neue\\",\\"Helvetica\\",\\"Arial\\",sans-serif"`;
